components/PlaybackTask.brs

```diff
@@ -0,0 +1,244 @@
+  end if
+
+  while(true)
+    msg = Wait(0, m.port)
+    msgType = type(msg)
+
+    ' ? "event type:" msgType
+    if msgType = "roSGNodeEvent"
+      field = msg.getField()
+      ? "roSGNodeEvent msg.getField()" field
+      if field = "position" then 
+        onPositionChanged(msg)
+      else if field = "event" then
+        onTrueXEvent(msg)
+      end if
+      ' TODO: Video complete event
+    end if
+  end while
+end sub
+
+sub checkPreroll()
+  ads = m.raf.getAds()
+  if ads = invalid then return
+
+  for each adPod in ads
+    if adPod.rendersequence = "preroll"
+      handleAds(adPod)
+      exit for
+    end if
+  end for
+end sub
+
+sub onPositionChanged(event)
+  position = event.getData()
+  ' ? "onPositionChanged: " position
+
+  ads = m.raf.getAds(event)
+  handleAds(ads)
+end sub
+
+sub onTruexEvent(event)
+  data = event.getData()
+  eventType = data.type
+
+  ? "TRUE[X] >>> PlaybackTask::onTrueXEvent(): " eventType
+
+  types = {
+    "ADSTARTED": "adStarted",
+    "ADCOMPLETED": "adCompleted"
+    "ADFIRSTQUARTILE": "adFirstQuartile",
+    "ADSECONDQUARTILE": "adSecondQuartile",
+    "ADTHIRDQUARTILE": "adThirdQuartile",
+    "ADPOSITION": "adPosition",
+    "ADERROR": "adError",
+    "ADFREEPOD": "adFreePod"
+    "NOADSAVAILABLE": "noAdsAvailable",
+    "OPTIN": "optIn",
+    "OPTOUT": "optOut",
+    "USERCANCEL": "userCancel",
+    "USERCANCELSTREAM": "userCancelStream",
+    "SKIPCARDSHOWN": "skipCardShown"
+  }
+
+  ' TODO: Various event formats
+  ' TODO: Combine statements into one if statement.  Or something else like SWITCH?
+
+  if eventType = types.SKIPCARDSHOWN
+  end if
+
+  if eventType = types.ADFREEPOD
+    m.skipAds = true
+  end if
+
+  if eventType = types.ADCOMPLETED
+    restorePlayback()
+  end if
+
+  if eventType = types.USERCANCEL OR eventType = types.OPTOUT OR eventType = types.NOADSAVAILABLE OR eventType = types.ADERROR
+    cleanUpAdRenderer()
+    restorePlayback()
+  end if
+
+  if eventType = types.USERCANCELSTREAM
+    cleanUpAdRenderer()
+    m.top.playbackEvent = { trigger: "cancelStream" } ' Might want to change this API since this is the flows API to the scene
+  end if
+end sub
+
+sub handleAds(ads)
+  if ads <> invalid AND ads.ads.count() > 0
+    firstAd = ads.ads[0] 'Assume truex can only be first ad
+
+    if m.skipAds then
+      ads.viewed = true
+      m.skipAds = false
+
+      ' Show choice card if needed?
+      return
+    end if
+
+
+    ' TODO: Figure out how to get metadata into the Roku ad parser.  AdParameters?
+    ' Hacky conditional for now
+    if firstAd.streams <> invalid AND firstAd.creativeid = "truex-test-id"
+      url = firstAd.streams[0].url
+      m.truexAd = firstAd
+      m.truexAd.renderSequence = ads.renderSequence
+
+      m.truexAd.params = {
+        vast_config_url: url,
+        placement_hash: "74fca63c733f098340b0a70489035d683024440d" 'Placeholder
+      }
+
+      ads.ads.delete(0) ' Removes it from future ad handling for raf
```

Comment:
Currently need to do this, otherwise for opt out type flows, RAF will attempt to play the TrueX ad with its own ad renderer, followed by the regular ads.  The desired flow is to ignore the TrueX ad, and have RAF play the other ads.

Crackle was doing this, and the reasons for it make sense.  But I do see your concern.  Replayability becomes a concern.  One way to handle this is to parse all the ads ahead of time, then introduce some kind of a mapping, so the information isn't lost.  I am of the opinion rewinding streams is relatively uncommon, in terms of replaying ads.  So this could be more of an edge case concern too.